File: templates/examples/core-components/mongodb/example-crud.py
#!/usr/bin/env python3
"""
MongoDB CRUD Example (Python)

This example demonstrates MongoDB CRUD operations using PyMongo
with proper error handling and connection management.
"""

import os
import sys
from typing import Any, Dict, List, Optional
from datetime import datetime
from dataclasses import dataclass, asdict
from pymongo import MongoClient, ASCENDING, DESCENDING
from pymongo.errors import ConnectionFailure, DuplicateKeyError, PyMongoError
from bson import ObjectId
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

@dataclass
class User:
    """User data model."""
    username: str
    email: str
    age: int
    role: str = "user"
    created_at: datetime = None
    updated_at: datetime = None
    _id: Optional[ObjectId] = None
    
    def __post_init__(self):
        if self.created_at is None:
            self.created_at = datetime.now()
        if self.updated_at is None:
            self.updated_at = datetime.now()

@dataclass
class Product:
    """Product data model."""
    name: str
    price: float
    category: str
    description: str = ""
    in_stock: bool = True
    tags: List[str] = None
    created_at: datetime = None
    _id: Optional[ObjectId] = None
    
    def __post_init__(self):
        if self.tags is None:
            self.tags = []
        if self.created_at is None:
            self.created_at = datetime.now()

class MongoDBClient:
    """MongoDB client with CRUD operations."""
    
    def __init__(self, connection_string: str = None, database_name: str = "spinbox_demo"):
        """Initialize MongoDB client."""
        
        # Get connection string from environment or use default
        self.connection_string = connection_string or os.getenv(
            "MONGODB_URI", 
            "mongodb://localhost:27017/"
        )
        
        self.database_name = database_name
        
        # Initialize client
        try:
            self.client = MongoClient(self.connection_string)
            self.db = self.client[self.database_name]
            
            # Test connection
            self.client.admin.command('ping')
            print(f"✅ Connected to MongoDB: {self.database_name}")
            
        except ConnectionFailure as e:
            print(f"❌ Failed to connect to MongoDB: {e}")
            raise
    
    def create_indexes(self):
        """Create indexes for better performance."""
        try:
            # Users collection indexes
            users_collection = self.db.users
            users_collection.create_index([("username", ASCENDING)], unique=True)
            users_collection.create_index([("email", ASCENDING)], unique=True)
            users_collection.create_index([("created_at", DESCENDING)])
            
            # Products collection indexes
            products_collection = self.db.products
            products_collection.create_index([("name", ASCENDING)])
            products_collection.create_index([("category", ASCENDING)])
            products_collection.create_index([("price", ASCENDING)])
            products_collection.create_index([("tags", ASCENDING)])
            
            print("✅ Indexes created successfully")
            
        except Exception as e:
            print(f"❌ Error creating indexes: {e}")
    
    def insert_user(self, user: User) -> Optional[ObjectId]:
        """Insert a new user."""
        try:
            user_dict = asdict(user)
            
            # Remove None _id for insertion
            if user_dict.get("_id") is None:
                user_dict.pop("_id", None)
            
            result = self.db.users.insert_one(user_dict)
            print(f"✅ User {user.username} inserted with ID: {result.inserted_id}")
            return result.inserted_id
            
        except DuplicateKeyError as e:
            print(f"❌ User {user.username} already exists: {e}")
            return None
        except Exception as e:
            print(f"❌ Error inserting user: {e}")
            return None
    
    def get_user(self, user_id: str = None, username: str = None) -> Optional[Dict[str, Any]]:
        """Get user by ID or username."""
        try:
            query = {}
            if user_id:
                query["_id"] = ObjectId(user_id)
            elif username:
                query["username"] = username
            else:
                return None
            
            user = self.db.users.find_one(query)
            
            if user:
                print(f"✅ Found user: {user.get('username')}")
            else:
                print("❌ User not found")
            
            return user
            
        except Exception as e:
            print(f"❌ Error getting user: {e}")
            return None
    
    def get_users(self, limit: int = 10, skip: int = 0, sort_by: str = "created_at") -> List[Dict[str, Any]]:
        """Get multiple users with pagination."""
        try:
            sort_direction = DESCENDING if sort_by == "created_at" else ASCENDING
            
            users = list(self.db.users.find()
                        .sort(sort_by, sort_direction)
                        .skip(skip)
                        .limit(limit))
            
            print(f"✅ Found {len(users)} users")
            return users
            
        except Exception as e:
            print(f"❌ Error getting users: {e}")
            return []
    
    def update_user(self, user_id: str, updates: Dict[str, Any]) -> bool:
        """Update user by ID."""
        try:
            # Add updated_at timestamp
            updates["updated_at"] = datetime.now()
            
            result = self.db.users.update_one(
                {"_id": ObjectId(user_id)},
                {"$set": updates}
            )
            
            if result.modified_count > 0:
                print(f"✅ User {user_id} updated successfully")
                return True
            else:
                print(f"❌ No user found with ID: {user_id}")
                return False
                
        except Exception as e:
            print(f"❌ Error updating user: {e}")
            return False
    
    def delete_user(self, user_id: str) -> bool:
        """Delete user by ID."""
        try:
            result = self.db.users.delete_one({"_id": ObjectId(user_id)})
            
            if result.deleted_count > 0:
                print(f"✅ User {user_id} deleted successfully")
                return True
            else:
                print(f"❌ No user found with ID: {user_id}")
                return False
                
        except Exception as e:
            print(f"❌ Error deleting user: {e}")
            return False
    
    def insert_product(self, product: Product) -> Optional[ObjectId]:
        """Insert a new product."""
        try:
            product_dict = asdict(product)
            
            # Remove None _id for insertion
            if product_dict.get("_id") is None:
                product_dict.pop("_id", None)
            
            result = self.db.products.insert_one(product_dict)
            print(f"✅ Product {product.name} inserted with ID: {result.inserted_id}")
            return result.inserted_id
            
        except Exception as e:
            print(f"❌ Error inserting product: {e}")
            return None
    
    def get_products(self, 
                    category: str = None, 
                    min_price: float = None, 
                    max_price: float = None,
                    in_stock: bool = None,
                    limit: int = 10) -> List[Dict[str, Any]]:
        """Get products with filtering."""
        try:
            query = {}
            
            # Build query filters
            if category:
                query["category"] = category
            if min_price is not None or max_price is not None:
                price_filter = {}
                if min_price is not None:
                    price_filter["$gte"] = min_price
                if max_price is not None:
                    price_filter["$lte"] = max_price
                query["price"] = price_filter
            if in_stock is not None:
                query["in_stock"] = in_stock
            
            products = list(self.db.products.find(query).limit(limit))
            
            print(f"✅ Found {len(products)} products")
            return products
            
        except Exception as e:
            print(f"❌ Error getting products: {e}")
            return []
    
    def search_products(self, search_term: str, limit: int = 10) -> List[Dict[str, Any]]:
        """Search products by name or description."""
        try:
            # Create text search query
            query = {
                "$or": [
                    {"name": {"$regex": search_term, "$options": "i"}},
                    {"description": {"$regex": search_term, "$options": "i"}},
                    {"tags": {"$regex": search_term, "$options": "i"}}
                ]
            }
            
            products = list(self.db.products.find(query).limit(limit))
            
            print(f"✅ Found {len(products)} products matching '{search_term}'")
            return products
            
        except Exception as e:
            print(f"❌ Error searching products: {e}")
            return []
    
    def get_categories(self) -> List[str]:
        """Get all product categories."""
        try:
            categories = self.db.products.distinct("category")
            print(f"✅ Found {len(categories)} categories")
            return categories
            
        except Exception as e:
            print(f"❌ Error getting categories: {e}")
            return []
    
    def get_collection_stats(self, collection_name: str) -> Dict[str, Any]:
        """Get statistics for a collection."""
        try:
            stats = self.db.command("collStats", collection_name)
            
            return {
                "count": stats.get("count", 0),
                "size": stats.get("size", 0),
                "avgObjSize": stats.get("avgObjSize", 0),
                "storageSize": stats.get("storageSize", 0),
                "indexes": stats.get("nindexes", 0)
            }
            
        except Exception as e:
            print(f"❌ Error getting collection stats: {e}")
            return {}
    
    def drop_collection(self, collection_name: str) -> bool:
        """Drop a collection."""
        try:
            self.db.drop_collection(collection_name)
            print(f"✅ Collection {collection_name} dropped")
            return True
            
        except Exception as e:
            print(f"❌ Error dropping collection: {e}")
            return False
    
    def close(self):
        """Close the MongoDB connection."""
        try:
            self.client.close()
            print("✅ MongoDB connection closed")
        except Exception as e:
            print(f"❌ Error closing connection: {e}")

class MongoDBDemo:
    """MongoDB CRUD operations demonstration."""
    
    def __init__(self):
        """Initialize the demo."""
        self.client = MongoDBClient()
        self.client.create_indexes()
    
    def seed_data(self):
        """Seed database with sample data."""
        print("\n🌱 Seeding sample data...")
        
        # Sample users
        users = [
            User(username="alice", email="alice@example.com", age=25, role="admin"),
            User(username="bob", email="bob@example.com", age=30, role="user"),
            User(username="charlie", email="charlie@example.com", age=35, role="user"),
            User(username="diana", email="diana@example.com", age=28, role="moderator")
        ]
        
        user_ids = []
        for user in users:
            user_id = self.client.insert_user(user)
            if user_id:
                user_ids.append(user_id)
        
        # Sample products
        products = [
            Product(name="Laptop", price=999.99, category="Electronics", 
                   description="High-performance laptop", tags=["computer", "work"]),
            Product(name="Coffee Mug", price=15.99, category="Kitchen", 
                   description="Ceramic coffee mug", tags=["coffee", "ceramic"]),
            Product(name="Python Book", price=39.99, category="Books", 
                   description="Learn Python programming", tags=["python", "programming"]),
            Product(name="Wireless Mouse", price=29.99, category="Electronics", 
                   description="Ergonomic wireless mouse", tags=["mouse", "wireless"]),
            Product(name="Notebook", price=8.99, category="Office", 
                   description="Lined notebook for writing", tags=["writing", "paper"])
        ]
        
        product_ids = []
        for product in products:
            product_id = self.client.insert_product(product)
            if product_id:
                product_ids.append(product_id)
        
        return user_ids, product_ids
    
    def demo_user_operations(self):
        """Demonstrate user CRUD operations."""
        print("\n👤 User Operations Demo")
        print("-" * 30)
        
        # Get all users
        users = self.client.get_users(limit=5)
        print(f"📋 Users in database: {len(users)}")
        
        if users:
            # Get first user
            first_user = users[0]
            user_id = str(first_user["_id"])
            
            # Update user
            updates = {"age": 26, "role": "super_admin"}
            self.client.update_user(user_id, updates)
            
            # Get updated user
            updated_user = self.client.get_user(user_id=user_id)
            if updated_user:
                print(f"📄 Updated user: {updated_user['username']} - Age: {updated_user['age']}")
    
    def demo_product_operations(self):
        """Demonstrate product CRUD operations."""
        print("\n🛍️ Product Operations Demo")
        print("-" * 30)
        
        # Get products by category
        electronics = self.client.get_products(category="Electronics")
        print(f"💻 Electronics products: {len(electronics)}")
        
        # Search products
        search_results = self.client.search_products("python")
        print(f"🔍 Search results for 'python': {len(search_results)}")
        
        # Get products by price range
        affordable_products = self.client.get_products(min_price=10, max_price=50)
        print(f"💰 Products $10-$50: {len(affordable_products)}")
        
        # Get all categories
        categories = self.client.get_categories()
        print(f"📂 Product categories: {', '.join(categories)}")
    
    def demo_aggregation(self):
        """Demonstrate aggregation operations."""
        print("\n📊 Aggregation Demo")
        print("-" * 30)
        
        # User statistics by role
        user_stats = list(self.client.db.users.aggregate([
            {"$group": {"_id": "$role", "count": {"$sum": 1}, "avg_age": {"$avg": "$age"}}},
            {"$sort": {"count": -1}}
        ]))
        
        print("👥 User statistics by role:")
        for stat in user_stats:
            print(f"  {stat['_id']}: {stat['count']} users, avg age: {stat['avg_age']:.1f}")
        
        # Product statistics by category
        product_stats = list(self.client.db.products.aggregate([
            {"$group": {
                "_id": "$category", 
                "count": {"$sum": 1}, 
                "avg_price": {"$avg": "$price"},
                "total_value": {"$sum": "$price"}
            }},
            {"$sort": {"total_value": -1}}
        ]))
        
        print("\n🛍️ Product statistics by category:")
        for stat in product_stats:
            print(f"  {stat['_id']}: {stat['count']} products, avg: ${stat['avg_price']:.2f}, total: ${stat['total_value']:.2f}")
    
    def demo_advanced_queries(self):
        """Demonstrate advanced query operations."""
        print("\n🔍 Advanced Queries Demo")
        print("-" * 30)
        
        # Find users created in the last hour
        recent_users = list(self.client.db.users.find({
            "created_at": {"$gte": datetime.now().replace(hour=0, minute=0, second=0)}
        }))
        print(f"🕐 Users created today: {len(recent_users)}")
        
        # Find products with specific tags
        tagged_products = list(self.client.db.products.find({
            "tags": {"$in": ["computer", "programming"]}
        }))
        print(f"🏷️ Products with 'computer' or 'programming' tags: {len(tagged_products)}")
        
        # Find products not in stock
        out_of_stock = list(self.client.db.products.find({
            "in_stock": False
        }))
        print(f"❌ Out of stock products: {len(out_of_stock)}")
    
    def cleanup(self):
        """Clean up demo data."""
        print("\n🧹 Cleaning up demo data...")
        
        # Drop collections
        self.client.drop_collection("users")
        self.client.drop_collection("products")
        
        # Close connection
        self.client.close()

def main():
    """Main function demonstrating MongoDB CRUD operations."""
    print("🍃 MongoDB CRUD Example (Python)")
    print("=" * 50)
    
    try:
        # Initialize demo
        demo = MongoDBDemo()
        
        # Seed data
        user_ids, product_ids = demo.seed_data()
        
        # Demo operations
        demo.demo_user_operations()
        demo.demo_product_operations()
        demo.demo_aggregation()
        demo.demo_advanced_queries()
        
        # Show collection stats
        print("\n📊 Collection Statistics")
        print("-" * 30)
        user_stats = demo.client.get_collection_stats("users")
        product_stats = demo.client.get_collection_stats("products")
        
        print(f"👥 Users: {user_stats.get('count', 0)} documents")
        print(f"🛍️ Products: {product_stats.get('count', 0)} documents")
        
        # Interactive mode
        print("\n" + "=" * 50)
        print("🎯 Interactive MongoDB Demo")
        print("Commands:")
        print("  'users' - List all users")
        print("  'products' - List all products")
        print("  'search <term>' - Search products")
        print("  'category <name>' - Get products by category")
        print("  'stats' - Show collection statistics")
        print("  'cleanup' - Clean up demo data")
        print("  'quit' - Exit")
        print("-" * 50)
        
        while True:
            user_input = input("\n🍃 MongoDB> ").strip()
            
            if user_input.lower() == 'quit':
                print("👋 Goodbye!")
                break
            elif user_input.lower() == 'users':
                users = demo.client.get_users(limit=10)
                print(f"👥 Users ({len(users)}):")
                for user in users:
                    print(f"  - {user['username']} ({user['email']}) - {user['role']}")
            elif user_input.lower() == 'products':
                products = demo.client.get_products(limit=10)
                print(f"🛍️ Products ({len(products)}):")
                for product in products:
                    print(f"  - {product['name']} - ${product['price']:.2f} ({product['category']})")
            elif user_input.startswith('search '):
                term = user_input.split(' ', 1)[1]
                products = demo.client.search_products(term)
                print(f"🔍 Search results for '{term}' ({len(products)}):")
                for product in products:
                    print(f"  - {product['name']} - ${product['price']:.2f}")
            elif user_input.startswith('category '):
                category = user_input.split(' ', 1)[1]
                products = demo.client.get_products(category=category)
                print(f"📂 Products in '{category}' ({len(products)}):")
                for product in products:
                    print(f"  - {product['name']} - ${product['price']:.2f}")
            elif user_input.lower() == 'stats':
                user_stats = demo.client.get_collection_stats("users")
                product_stats = demo.client.get_collection_stats("products")
                print("📊 Collection Statistics:")
                print(f"  Users: {user_stats.get('count', 0)} documents")
                print(f"  Products: {product_stats.get('count', 0)} documents")
            elif user_input.lower() == 'cleanup':
                demo.cleanup()
                print("✅ Demo data cleaned up")
                break
            elif user_input:
                print("Unknown command. Type 'quit' to exit.")
        
        # Final cleanup
        if user_input.lower() != 'cleanup':
            demo.cleanup()
        
    except KeyboardInterrupt:
        print("\n\n👋 Demo interrupted by user")
    except Exception as e:
        print(f"\n❌ Unexpected error: {str(e)}")
        sys.exit(1)

if __name__ == "__main__":
    main()

File: templates/examples/core-components/mongodb/example-crud.js
/**
 * MongoDB CRUD Examples
 * 
 * This example demonstrates:
 * - Basic CRUD operations
 * - Document validation
 * - Query optimization
 * - Bulk operations
 * - Error handling
 * 
 * Usage:
 * npm install mongodb
 * node example-crud.js
 */

const { MongoClient, ObjectId } = require('mongodb');

// MongoDB connection configuration
const MONGODB_URI = process.env.MONGODB_URI || 'mongodb://localhost:27017';
const DATABASE_NAME = 'spinbox_examples';

// ==========================================
// DATABASE CONNECTION
// ==========================================

class DatabaseManager {
    constructor(uri, dbName) {
        this.uri = uri;
        this.dbName = dbName;
        this.client = null;
        this.db = null;
    }

    async connect() {
        try {
            this.client = new MongoClient(this.uri, {
                useNewUrlParser: true,
                useUnifiedTopology: true,
                maxPoolSize: 10,
                serverSelectionTimeoutMS: 5000,
                socketTimeoutMS: 45000,
            });

            await this.client.connect();
            this.db = this.client.db(this.dbName);
            
            console.log(`✅ Connected to MongoDB: ${this.dbName}`);
            return this.db;
        } catch (error) {
            console.error('❌ MongoDB connection error:', error);
            throw error;
        }
    }

    async disconnect() {
        if (this.client) {
            await this.client.close();
            console.log('🔌 Disconnected from MongoDB');
        }
    }

    getCollection(name) {
        if (!this.db) {
            throw new Error('Database not connected');
        }
        return this.db.collection(name);
    }
}

// ==========================================
// USER MANAGEMENT SYSTEM
// ==========================================

class UserManager {
    constructor(database) {
        this.db = database;
        this.collection = database.getCollection('users');
        this.setupValidation();
    }

    /**
     * Set up document validation
     */
    async setupValidation() {
        try {
            await this.db.db.createCollection('users', {
                validator: {
                    $jsonSchema: {
                        bsonType: 'object',
                        required: ['email', 'username', 'createdAt'],
                        properties: {
                            email: {
                                bsonType: 'string',
                                pattern: '^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\\.[a-zA-Z]{2,}$',
                                description: 'Valid email address is required'
                            },
                            username: {
                                bsonType: 'string',
                                minLength: 3,
                                maxLength: 50,
                                description: 'Username must be 3-50 characters'
                            },
                            age: {
                                bsonType: 'int',
                                minimum: 0,
                                maximum: 150,
                                description: 'Age must be between 0 and 150'
                            },
                            status: {
                                bsonType: 'string',
                                enum: ['active', 'inactive', 'pending', 'banned'],
                                description: 'Status must be one of: active, inactive, pending, banned'
                            },
                            createdAt: {
                                bsonType: 'date',
                                description: 'Creation date is required'
                            }
                        }
                    }
                }
            });
        } catch (error) {
            // Collection might already exist
            if (error.code !== 48) { // NamespaceExists
                console.error('Validation setup error:', error);
            }
        }
    }

    // ==========================================
    // CREATE OPERATIONS
    // ==========================================

    /**
     * Create a new user
     */
    async createUser(userData) {
        try {
            const user = {
                ...userData,
                createdAt: new Date(),
                updatedAt: new Date(),
                status: userData.status || 'active'
            };

            const result = await this.collection.insertOne(user);
            console.log(`✅ User created with ID: ${result.insertedId}`);
            
            return {
                success: true,
                id: result.insertedId,
                user: { ...user, _id: result.insertedId }
            };
        } catch (error) {
            console.error('❌ Error creating user:', error);
            return { success: false, error: error.message };
        }
    }

    /**
     * Create multiple users
     */
    async createUsers(usersData) {
        try {
            const users = usersData.map(userData => ({
                ...userData,
                createdAt: new Date(),
                updatedAt: new Date(),
                status: userData.status || 'active'
            }));

            const result = await this.collection.insertMany(users);
            console.log(`✅ Created ${result.insertedCount} users`);
            
            return {
                success: true,
                count: result.insertedCount,
                ids: result.insertedIds
            };
        } catch (error) {
            console.error('❌ Error creating users:', error);
            return { success: false, error: error.message };
        }
    }

    // ==========================================
    // READ OPERATIONS
    // ==========================================

    /**
     * Get user by ID
     */
    async getUserById(id) {
        try {
            const objectId = ObjectId.isValid(id) ? new ObjectId(id) : id;
            const user = await this.collection.findOne({ _id: objectId });
            
            if (!user) {
                return { success: false, error: 'User not found' };
            }
            
            return { success: true, user };
        } catch (error) {
            console.error('❌ Error getting user:', error);
            return { success: false, error: error.message };
        }
    }

    /**
     * Get user by email
     */
    async getUserByEmail(email) {
        try {
            const user = await this.collection.findOne({ email });
            
            if (!user) {
                return { success: false, error: 'User not found' };
            }
            
            return { success: true, user };
        } catch (error) {
            console.error('❌ Error getting user by email:', error);
            return { success: false, error: error.message };
        }
    }

    /**
     * Get all users with pagination
     */
    async getUsers(options = {}) {
        try {
            const {
                page = 1,
                limit = 10,
                sortBy = 'createdAt',
                sortOrder = -1,
                filter = {},
                projection = {}
            } = options;

            const skip = (page - 1) * limit;
            
            const [users, total] = await Promise.all([
                this.collection
                    .find(filter, { projection })
                    .sort({ [sortBy]: sortOrder })
                    .skip(skip)
                    .limit(limit)
                    .toArray(),
                this.collection.countDocuments(filter)
            ]);

            return {
                success: true,
                users,
                pagination: {
                    page,
                    limit,
                    total,
                    pages: Math.ceil(total / limit),
                    hasNext: page < Math.ceil(total / limit),
                    hasPrev: page > 1
                }
            };
        } catch (error) {
            console.error('❌ Error getting users:', error);
            return { success: false, error: error.message };
        }
    }

    /**
     * Search users
     */
    async searchUsers(query, options = {}) {
        try {
            const { limit = 10, sortBy = 'createdAt', sortOrder = -1 } = options;

            const filter = {
                $or: [
                    { username: { $regex: query, $options: 'i' } },
                    { email: { $regex: query, $options: 'i' } },
                    { firstName: { $regex: query, $options: 'i' } },
                    { lastName: { $regex: query, $options: 'i' } }
                ]
            };

            const users = await this.collection
                .find(filter)
                .sort({ [sortBy]: sortOrder })
                .limit(limit)
                .toArray();

            return { success: true, users, count: users.length };
        } catch (error) {
            console.error('❌ Error searching users:', error);
            return { success: false, error: error.message };
        }
    }

    // ==========================================
    // UPDATE OPERATIONS
    // ==========================================

    /**
     * Update user by ID
     */
    async updateUser(id, updates) {
        try {
            const objectId = ObjectId.isValid(id) ? new ObjectId(id) : id;
            
            // Add updatedAt timestamp
            const updateData = {
                ...updates,
                updatedAt: new Date()
            };

            const result = await this.collection.updateOne(
                { _id: objectId },
                { $set: updateData }
            );

            if (result.matchedCount === 0) {
                return { success: false, error: 'User not found' };
            }

            console.log(`✅ User updated: ${id}`);
            return { success: true, modifiedCount: result.modifiedCount };
        } catch (error) {
            console.error('❌ Error updating user:', error);
            return { success: false, error: error.message };
        }
    }

    /**
     * Update multiple users
     */
    async updateUsers(filter, updates) {
        try {
            const updateData = {
                ...updates,
                updatedAt: new Date()
            };

            const result = await this.collection.updateMany(
                filter,
                { $set: updateData }
            );

            console.log(`✅ Updated ${result.modifiedCount} users`);
            return { success: true, modifiedCount: result.modifiedCount };
        } catch (error) {
            console.error('❌ Error updating users:', error);
            return { success: false, error: error.message };
        }
    }

    /**
     * Increment user field
     */
    async incrementUserField(id, field, amount = 1) {
        try {
            const objectId = ObjectId.isValid(id) ? new ObjectId(id) : id;
            
            const result = await this.collection.updateOne(
                { _id: objectId },
                { 
                    $inc: { [field]: amount },
                    $set: { updatedAt: new Date() }
                }
            );

            if (result.matchedCount === 0) {
                return { success: false, error: 'User not found' };
            }

            return { success: true, modifiedCount: result.modifiedCount };
        } catch (error) {
            console.error('❌ Error incrementing field:', error);
            return { success: false, error: error.message };
        }
    }

    // ==========================================
    // DELETE OPERATIONS
    // ==========================================

    /**
     * Delete user by ID
     */
    async deleteUser(id) {
        try {
            const objectId = ObjectId.isValid(id) ? new ObjectId(id) : id;
            
            const result = await this.collection.deleteOne({ _id: objectId });

            if (result.deletedCount === 0) {
                return { success: false, error: 'User not found' };
            }

            console.log(`✅ User deleted: ${id}`);
            return { success: true, deletedCount: result.deletedCount };
        } catch (error) {
            console.error('❌ Error deleting user:', error);
            return { success: false, error: error.message };
        }
    }

    /**
     * Delete multiple users
     */
    async deleteUsers(filter) {
        try {
            const result = await this.collection.deleteMany(filter);
            
            console.log(`✅ Deleted ${result.deletedCount} users`);
            return { success: true, deletedCount: result.deletedCount };
        } catch (error) {
            console.error('❌ Error deleting users:', error);
            return { success: false, error: error.message };
        }
    }

    /**
     * Soft delete user (mark as inactive)
     */
    async softDeleteUser(id) {
        try {
            const result = await this.updateUser(id, { 
                status: 'inactive',
                deletedAt: new Date()
            });
            
            if (result.success) {
                console.log(`✅ User soft deleted: ${id}`);
            }
            
            return result;
        } catch (error) {
            console.error('❌ Error soft deleting user:', error);
            return { success: false, error: error.message };
        }
    }

    // ==========================================
    // BULK OPERATIONS
    // ==========================================

    /**
     * Bulk operations
     */
    async bulkWrite(operations) {
        try {
            const result = await this.collection.bulkWrite(operations);
            
            console.log(`✅ Bulk operation completed:`, {
                inserted: result.insertedCount,
                modified: result.modifiedCount,
                deleted: result.deletedCount
            });
            
            return { success: true, result };
        } catch (error) {
            console.error('❌ Error in bulk operation:', error);
            return { success: false, error: error.message };
        }
    }

    // ==========================================
    // UTILITY METHODS
    // ==========================================

    /**
     * Get user statistics
     */
    async getUserStats() {
        try {
            const stats = await this.collection.aggregate([
                {
                    $group: {
                        _id: '$status',
                        count: { $sum: 1 }
                    }
                }
            ]).toArray();

            const total = await this.collection.countDocuments();
            
            return {
                success: true,
                stats: {
                    total,
                    byStatus: stats.reduce((acc, item) => {
                        acc[item._id] = item.count;
                        return acc;
                    }, {})
                }
            };
        } catch (error) {
            console.error('❌ Error getting user stats:', error);
            return { success: false, error: error.message };
        }
    }

    /**
     * Get recent users
     */
    async getRecentUsers(limit = 10) {
        try {
            const users = await this.collection
                .find({})
                .sort({ createdAt: -1 })
                .limit(limit)
                .toArray();

            return { success: true, users };
        } catch (error) {
            console.error('❌ Error getting recent users:', error);
            return { success: false, error: error.message };
        }
    }

    /**
     * Check if user exists
     */
    async userExists(email) {
        try {
            const count = await this.collection.countDocuments({ email });
            return count > 0;
        } catch (error) {
            console.error('❌ Error checking user existence:', error);
            return false;
        }
    }
}

// ==========================================
// EXAMPLE USAGE
// ==========================================

async function runCRUDExamples() {
    const dbManager = new DatabaseManager(MONGODB_URI, DATABASE_NAME);
    
    try {
        // Connect to database
        await dbManager.connect();
        const userManager = new UserManager(dbManager);

        console.log('=== CREATE OPERATIONS ===');
        
        // Create single user
        const createResult = await userManager.createUser({
            email: 'john.doe@example.com',
            username: 'johndoe',
            firstName: 'John',
            lastName: 'Doe',
            age: 30
        });
        
        const userId = createResult.id;
        console.log('Created user:', createResult);

        // Create multiple users
        const bulkCreateResult = await userManager.createUsers([
            {
                email: 'jane.smith@example.com',
                username: 'janesmith',
                firstName: 'Jane',
                lastName: 'Smith',
                age: 28
            },
            {
                email: 'bob.wilson@example.com',
                username: 'bobwilson',
                firstName: 'Bob',
                lastName: 'Wilson',
                age: 35
            }
        ]);
        
        console.log('Bulk create result:', bulkCreateResult);

        console.log('\\n=== READ OPERATIONS ===');
        
        // Get user by ID
        const getUserResult = await userManager.getUserById(userId);
        console.log('Get user by ID:', getUserResult);

        // Get user by email
        const getUserByEmailResult = await userManager.getUserByEmail('jane.smith@example.com');
        console.log('Get user by email:', getUserByEmailResult);

        // Get all users with pagination
        const getUsersResult = await userManager.getUsers({
            page: 1,
            limit: 5,
            sortBy: 'createdAt',
            sortOrder: -1
        });
        console.log('Get users:', getUsersResult);

        // Search users
        const searchResult = await userManager.searchUsers('john');
        console.log('Search users:', searchResult);

        console.log('\\n=== UPDATE OPERATIONS ===');
        
        // Update user
        const updateResult = await userManager.updateUser(userId, {
            firstName: 'John Updated',
            age: 31
        });
        console.log('Update user:', updateResult);

        // Update multiple users
        const bulkUpdateResult = await userManager.updateUsers(
            { status: 'active' },
            { lastActive: new Date() }
        );
        console.log('Bulk update result:', bulkUpdateResult);

        // Increment field
        const incrementResult = await userManager.incrementUserField(userId, 'loginCount', 1);
        console.log('Increment result:', incrementResult);

        console.log('\\n=== BULK OPERATIONS ===');
        
        // Bulk operations
        const bulkOps = [
            {
                updateOne: {
                    filter: { email: 'john.doe@example.com' },
                    update: { $set: { status: 'premium' } }
                }
            },
            {
                updateMany: {
                    filter: { age: { $gte: 30 } },
                    update: { $set: { category: 'senior' } }
                }
            }
        ];
        
        const bulkResult = await userManager.bulkWrite(bulkOps);
        console.log('Bulk write result:', bulkResult);

        console.log('\\n=== UTILITY OPERATIONS ===');
        
        // Get user statistics
        const statsResult = await userManager.getUserStats();
        console.log('User stats:', statsResult);

        // Get recent users
        const recentUsers = await userManager.getRecentUsers(3);
        console.log('Recent users:', recentUsers);

        // Check if user exists
        const userExists = await userManager.userExists('john.doe@example.com');
        console.log('User exists:', userExists);

        console.log('\\n=== DELETE OPERATIONS ===');
        
        // Soft delete
        const softDeleteResult = await userManager.softDeleteUser(userId);
        console.log('Soft delete result:', softDeleteResult);

        // Delete users by filter
        const deleteResult = await userManager.deleteUsers({
            status: 'inactive'
        });
        console.log('Delete result:', deleteResult);

        // Clean up remaining test data
        await userManager.deleteUsers({});
        console.log('Cleaned up test data');

    } catch (error) {
        console.error('Error in CRUD examples:', error);
    } finally {
        await dbManager.disconnect();
    }
}

// Run examples when script is executed directly
if (require.main === module) {
    runCRUDExamples().catch(console.error);
}

module.exports = {
    DatabaseManager,
    UserManager
};